### ISSUE TITLE
Downloads repeatedly stuck

### ISSUE THREAD

**Issue by kokoropop** [Otiel/BandcampDownloader#225]
App downloads only 0,15-0,45 Mb and stops. It looks like a problem with network, but network works good.

2022-07-28 09:49:31.0899  INFO   Starting download...
2022-07-28 09:49:31.1619  INFO   Retrieving album data for http://outubrometal.bandcamp.com/album/s
2022-07-28 09:49:31.2169  DEBUG  Downloading album info from url: http://outubrometal.bandcamp.com/album/s
2022-07-28 09:49:33.5230  INFO   Computing size for album "so"...
2022-07-28 09:49:33.7320  DEBUG  Retrieved the size of the MP3 file for the track "so I"
2022-07-28 09:49:33.7890  DEBUG  Retrieved the size of the MP3 file for the track "so II"
2022-07-28 09:49:33.8350  DEBUG  Downloading track "so I" from url: http://t4.bcbits.com/stream/5c3b8bce95a77672bc00278c87aa4bca/mp3-128/847891396?p=0&ts=1659080971&t=7ede2d73e06e119030845db1997248aabe10b002&token=1659080971_61d8a339e49ab66b06e5705a718546da508351f4
2022-07-28 09:49:33.8800  DEBUG  Downloading track "so I" from url: http://t4.bcbits.com/stream/5c3b8bce95a77672bc00278c87aa4bca/mp3-128/847891396?p=0&ts=1659080971&t=7ede2d73e06e119030845db1997248aabe10b002&token=1659080971_61d8a339e49ab66b06e5705a718546da508351f4
2022-07-28 09:49:33.9351  DEBUG  Downloading track "so II" from url: http://t4.bcbits.com/stream/bd6c5212ea40149e0c7d5cc736d3c797/mp3-128/3386989173?p=0&ts=1659080971&t=8262220db7e010320d02f187a0cd0fa3d38175e9&token=1659080971_4db36ec968a1dad018134c10331f8bbfb61043c8
2022-07-28 09:49:33.9791  DEBUG  Downloading track "so II" from url: http://t4.bcbits.com/stream/bd6c5212ea40149e0c7d5cc736d3c797/mp3-128/3386989173?p=0&ts=1659080971&t=8262220db7e010320d02f187a0cd0fa3d38175e9&token=1659080971_4db36ec968a1dad018134c10331f8bbfb61043c8
2022-07-28 09:49:42.6876  INFO   Downloads cancelled by user
![image](https://user-images.githubusercontent.com/71778730/181455230-cf567fd3-6a7e-4ca0-be7f-88e1c0b0c349.png)

**Comment by Otiel** [comment]
Can you reach http://t4.bcbits.com/stream/bd6c5212ea40149e0c7d5cc736d3c797/mp3-128/3386989173?p=0&ts=1659081457&t=b8d9b9fab3c59a82fcb2f0d17550003ba12c6195&token=1659081457_49444f4abfbdb08b225c34912dc2041f1f1d444e ?

Any BandcampDownloader entry in Event Viewer > Windows Logs > Application ?

**Comment by kokoropop** [comment]
> Can you reach http://t4.bcbits.com/stream/bd6c5212ea40149e0c7d5cc736d3c797/mp3-128/3386989173?p=0&ts=1659081457&t=b8d9b9fab3c59a82fcb2f0d17550003ba12c6195&token=1659081457_49444f4abfbdb08b225c34912dc2041f1f1d444e in your browser ?
> 
> Any BandcampDownloader entry in Event Viewer > Windows Logs > Application ?

Yes, I can reach it and download manually.

**Comment by Otiel** [comment]
Does this happen for all albums?

**Comment by kokoropop** [comment]
> Does this happen for all albums?

Yes. But one strange thing happened. Today's night at 1 a.m (in my local time) app worked, and I downloaded albums without problem. But now 9 a.m and app doesn't work again.

**Comment by Otiel** [comment]
It looks like your downloads are being restricted, most probably by Bandcamp.

**Comment by kokoropop** [comment]
> It looks like your downloads are being restricted, most probably by Bandcamp.

Is it possible to fix?

**Comment by Otiel** [comment]
It can happen if you download lots of albums. In that case, what works usually is to wait some time before downloading again.

**Comment by kokoropop** [comment]
I don't understand how Bandcamp restricts my downloads. What principles? Why I can download manually but can't do it with app? I use this app for five years, I usually download 40-45 tracks per day, not more. Why didn't I get restrictions all this time? I need this app every day. I can't wait some time.

**Comment by jtayl711** [comment]
I'm having this issue as well.  Is there simply a way to start the process over again and automatically skip items you've already downloaded?  I thought the program was doing this but then I noticed after one of my download sessions froze, that the size of the destination folder actually went down from the previous attempt.  This leads me to think it's actually erasing everything and starting over with each attempt??

**Comment by Otiel** [comment]
> Is there simply a way to start the process over again and automatically skip items you've already downloaded? I thought the program was doing this

Yes it should be skipping tracks already downloaded, if they have the same name (meaning you haven't the `filename format` in the meantime). Check as well that you have `Allowed file size difference` to 5%:

![image](https://user-images.githubusercontent.com/3851902/230603146-d054dd3c-b48b-4e6f-b1d8-0095ab38a1f9.png)

Did you notice some particular tracks that are being downloaded again when they should be skipped?

**Comment by jtayl711** [comment]
> > Is there simply a way to start the process over again and automatically skip items you've already downloaded? I thought the program was doing this
> 
> Yes it should be skipping tracks already downloaded, if they have the same name (meaning you haven't the `filename format` in the meantime). Check as well that you have `Allowed file size difference` to 5%:
> 
> ![image](https://user-images.githubusercontent.com/3851902/230603146-d054dd3c-b48b-4e6f-b1d8-0095ab38a1f9.png)
> 
> Did you notice some particular tracks that are being downloaded again when they should be skipped?

[BandcampDownloader.log](https://github.com/Otiel/BandcampDownloader/files/11179080/BandcampDownloader.log)
I didn't notice that explicitly.  Attaching my log file.  Do you think I'm experiencing the same problem as the above user?  Anyworkaround?

**Comment by Otiel** [comment]
1. Yes, I think Bandcamp is throttling you in some way, which causes some downloads to fail.
2. **But**, tracks already downloaded shouldn't be erased. It seems there is a bug here, I should look into it.

**Comment by jtayl711** [comment]
> 1. Yes, I think Bandcamp is throttling you in some way, which causes some downloads to fail.
> 2. **But**, tracks already downloaded shouldn't be erased. It seems there is a bug here, I should look into it.

I realized maybe it was because I unchecked the option to check size before downloads??  Still, if there's a workaround for being throttled please let me know.

**Comment by Otiel** [comment]
>I realized maybe it was because I unchecked the option to check size before downloads

Oh yes, my bad, I missed that! If file size isn't known before download, the check can't be done. So make sure this is checked:

![image](https://user-images.githubusercontent.com/3851902/230686570-2e81b22d-dc56-4627-9d14-12e1e543627c.png)

>if there's a workaround for being throttled please let me know.

Unfortunately, none that I know of. ☹️

**Comment by jtayl711** [comment]
> * > I realized maybe it was because I unchecked the option to check size before downloads
>   
>   
>   Oh yes, my bad, I missed that! If file size isn't known before download, the check can't be done. So make sure this is checked:
>   ![image](https://user-images.githubusercontent.com/3851902/230686570-2e81b22d-dc56-4627-9d14-12e1e543627c.png)
>   When track downloads are skipped, you should see the following log:
>   `Track already exists within allowed file size range`
> * > if there's a workaround for being throttled please let me know.
>   
>   
>   Unfortunately, none that I know of. ☹️

Any idea what it is that Bandcamp is looking for as an indication to begin throttling?  Is it number of downloads within a certain period of time?  If so, perhaps you can implement some kind of "dwell time" for which the program stops downloading?

**Comment by Otiel** [comment]
> Any idea what it is that Bandcamp is looking for as an indication to begin throttling?

Nothing certain. From what I've personally seen, it's the amount of downloads in a certain time. But people's experiences vary. Unless we could have a look at Bandcamp, we can't be sure. 

>  If so, perhaps you can implement some kind of "dwell time" for which the program stops downloading?

There's already a cool down mechanism implemented where the program waits before retrying a failed download. But without really knowing how Bandcamp throttles, we can't implement something perfect unfortunately.

**Comment by sequencerr** [comment]
I have the same issue. But it's downloading 1 album by 1 as intended, only struggles with full discography download. Maybe add some functionality to no download them all at once (one more checkbox when downloading discography)? Or monitor download speed and download progress and repeat from there?
@Otiel

**Comment by Otiel** [comment]
>But it's downloading 1 album by 1 as intended, only struggles with full discography download. Maybe add some functionality to no download them all at once (one more checkbox when downloading discography)?

If you checked "Settings → Downloads → Download one album at a time", it should work also for discographies.

>Because for the huge albums like https://adrianvonziegler.bandcamp.com/album/the-complete-discography I got throttled

Thanks for the example, never found such a big album. 😄 
I tried downloading it and I had no issue, so it's probably again a case of Bandcamp throttling depending on your history.

As stated in this issue, if you have those settings:
* Downloads → Retrieve file size before downloading tracks
* Advanced settings → Allowed file size difference : 5%

... then we you start downloading again the same album, it should not download again tracks that have been fully downloaded previously.

**Comment by Otiel** [comment]
Please try again with [v1.5.0](https://github.com/Otiel/BandcampDownloader/releases/tag/v1.5.0)

You can also lower the value of the new setting "Max concurrent tracks downloads" in order to limit the occurrences of the Bandcamp anti-spam blocking you:

<img width="606" height="372" alt="Image" src="https://github.com/user-attachments/assets/9d0d17f7-84a3-49b8-8b95-5ed4be3059ca" />
